cmd/buildkitd/timeout_interceptor.go

```diff
@@ -0,0 +1,93 @@
+package main
+
+// The sessionTimeout interceptors in this file are earthly-specific.
+// They are used to automatically cancel builds in CI or Satellites that have run for too long.
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+var sessionTimeout time.Duration
+
+func init() {
+	env, ok := os.LookupEnv("BUILDKIT_SESSION_TIMEOUT")
+	if !ok {
+		return
+	}
+	var err error
+	sessionTimeout, err = time.ParseDuration(env)
+	if err != nil {
+		panic(fmt.Sprintf("invalid value for 'BUILDKIT_SESSION_TIMEOUT': %s", env))
+	}
+}
+
+func unaryTimeoutInterceptor() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		if sessionTimeout > 0 {
+			ctx, cancel := context.WithTimeout(ctx, sessionTimeout)
+			defer cancel()
+			resp, err := handler(ctx, req)
+			if errors.Is(err, context.DeadlineExceeded) {
```

Comment:
here's a better example of what I mean by `// replace this with a select that will early exit when the parent function returns to avoid sleeping for 2 hours which would eat up a lot of thread resources.`
```
select {
    case <-endSignal:
        return
    case <-time.After(time.Millisecond):
        cancel(errSessionTimeout)
}
```

Reply:
alternatively a custom context might be possible here (which would avoid the thread) -- but given that this is a work-around until go 1.21 is released -- it might be more work than required.

Reply:
Yeah good point, there would be other reasons for a DeadlineExceeded error.. Let me see what I can do here.

Reply:
Added some code along the lines of your suggestion. Seems to be working.